File: HexaFlipTests/Source/UI/Screens/JCSFlipUIMainMenuScreenTest.m
//
//  JCSFlipUIMainMenuScreenTest.m
//  HexaFlip
//
//  Created by Christian Schuster on 05.11.13.
//  Copyright (c) 2013 Christian Schuster. All rights reserved.
//

#import "JCSFlipUIMainMenuScreen.h"
#import "JCSFlipGameCenterManager.h"
#import "JCSButton.h"
#import "JCSFlipUIEvents.h"

#import "OCMock.h"

@interface JCSFlipUIMainMenuScreen ()

// make selector visible to test
- (void)playerAuthenticationDidChange:(NSNotification *)notification;

@end

@interface JCSFlipUIMainMenuScreenTest : SenTestCase
@end

@implementation JCSFlipUIMainMenuScreenTest {
    JCSFlipUIMainMenuScreen *_underTest;
    id _gameCenterManagerMock;
    id _observerMock;
}

- (void)setUp {
    [super setUp];
    _underTest = [[JCSFlipUIMainMenuScreen alloc] init];
    _gameCenterManagerMock = [OCMockObject niceMockForClass:[JCSFlipGameCenterManager class]];
    [[[_gameCenterManagerMock stub] andReturn:_gameCenterManagerMock] sharedInstance];
    _observerMock = [OCMockObject observerMock];
    [[NSNotificationCenter defaultCenter] addMockObserver:_observerMock name:JCS_FLIP_UI_PLAY_SINGLE_EVENT_NAME object:nil];
    [[NSNotificationCenter defaultCenter] addMockObserver:_observerMock name:JCS_FLIP_UI_PLAY_MULTI_EVENT_NAME object:nil];
}

- (void)tearDown {
    NSNotificationCenter *nc = [NSNotificationCenter defaultCenter];
    [nc removeObserver:_observerMock];
    [super tearDown];
}

- (JCSButton *)playSingleButtonForScreen:(JCSFlipUIMainMenuScreen *)screen {
    CCMenu *menu = [screen.children objectAtIndex:0];
    return [menu.children objectAtIndex:0];
}

- (JCSButton *)playMultiButtonForScreen:(JCSFlipUIMainMenuScreen *)screen {
    // TODO use tags to identify children
    CCMenu *menu = [screen.children objectAtIndex:0];
    return [menu.children objectAtIndex:1];
}

- (void)testEventTriggeredWhenPlaySingleButtonActivated {
    [[_observerMock expect] notificationWithName:JCS_FLIP_UI_PLAY_SINGLE_EVENT_NAME object:_underTest];

    [[self playSingleButtonForScreen:_underTest] activate];
    
    [_observerMock verify];
}

- (void)testPlayMultiButtonDisabledWhenGameCenterNotAuthenticated {
    [_underTest performSelector:@selector(playerAuthenticationDidChange:) withObject:nil];
    
    STAssertFalse([self playMultiButtonForScreen:_underTest].isEnabled, @"button must be disabled");
}

- (void)testPlayMultiButtonEnabledWhenGameCenterAuthenticated {
    [[[_gameCenterManagerMock stub] andReturnValue:@YES] isLocalPlayerAuthenticated];
    
    [_underTest performSelector:@selector(playerAuthenticationDidChange:) withObject:nil];
    
    STAssertTrue([self playMultiButtonForScreen:_underTest].isEnabled, @"button must be enabled");
}

- (void)testEventTriggeredWhenPlayMultiButtonActivated {
    [[[_gameCenterManagerMock stub] andReturnValue:@YES] isLocalPlayerAuthenticated];
    [[_observerMock expect] notificationWithName:JCS_FLIP_UI_PLAY_MULTI_EVENT_NAME object:_underTest];

    [[self playMultiButtonForScreen:_underTest] activate];

    [_observerMock verify];
}

- (void)testObserverAddedWhenScreenEnabled {
    [[_gameCenterManagerMock expect] addPlayerAuthenticationObserver:_underTest selector:@selector(playerAuthenticationDidChange:)];
    
    [_underTest willActivateScreen];
    
    [_gameCenterManagerMock verify];
}

- (void)testObserverRemovedWhenScreenDisabled {
    [[_gameCenterManagerMock expect] removePlayerAuthenticationObserver:_underTest];
    
    [_underTest didDeactivateScreen];
    
    [_gameCenterManagerMock verify];
}

@end

File: HexaFlipTests/Source/UI/Screens/JCSFlipUIMainMenuScreenTest.swift
//
//  JCSFlipUIMainMenuScreenTest.swift
//  HexaFlip
//
//  Created by Christian Schuster on 04.11.14.
//  Copyright (c) 2014 Christian Schuster. All rights reserved.
//

import Foundation
import GameKit
import XCTest

// extension that exposes internals to the test (overridden by "ForTest" subclass)
extension JCSFlipUIMainMenuScreen {
    
    func isLocalPlayerAuthenticated() -> Bool {
        return false
    }
}

class JCSFlipUIMainMenuScreenTest : XCTestCase {
    
    let underTest = JCSFlipUIMainMenuScreenForTest()
    var events: [NSNotification] = []

    override func setUp() {
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "eventTriggered:", name: nil, object: nil)
    }
    
    override func tearDown() {
        NSNotificationCenter.defaultCenter().removeObserver(self)
    }
    
    func testEventTriggeredWhenPlaySingleButtonActivated() {
        playSingleButtonForScreen(underTest).activate()
        XCTAssertEqual(["JCS_FLIP_UI_PLAY_SINGLE_EVENT_NAME"], events.map { notification in notification.name })
    }

    func testEventTriggeredWhenPlayMultiButtonActivated() {
        playMultiButtonForScreen(underTest).activate()
        XCTAssertEqual(["JCS_FLIP_UI_PLAY_MULTI_EVENT_NAME"], events.map { notification in notification.name })
    }
    
    func testPlayMultiButtonState() {
        // activate screen while player is not authenticated
        underTest.authenticationStatus = false
        underTest.willActivateScreen()
        XCTAssertFalse(playMultiButtonForScreen(underTest).isEnabled(), "button must be initially disabled if player is not authenticated")

        // authenticate player while screen is active
        underTest.authenticationStatus = true
        XCTAssertTrue(playMultiButtonForScreen(underTest).isEnabled(), "button must be enabled when player is authenticated")
        
        // deauthenticate player while screen is active
        underTest.authenticationStatus = false
        XCTAssertFalse(playMultiButtonForScreen(underTest).isEnabled(), "button must stay disabled when player is deauthenticated")
 
        // authenticate player while screen is inactive
        underTest.didDeactivateScreen()
        underTest.authenticationStatus = true
        XCTAssertFalse(playMultiButtonForScreen(underTest).isEnabled(), "button must remain unchanged after screen has been deactivated")

        // activate screen while player is authenticated
        underTest.willActivateScreen()
        XCTAssertTrue(playMultiButtonForScreen(underTest).isEnabled(), "button must be initially enabled if player is authenticated")

        // deauthenticate player while screen is inactive
        underTest.didDeactivateScreen()
        underTest.authenticationStatus = false
        XCTAssertTrue(playMultiButtonForScreen(underTest).isEnabled(), "button must remain unchanged after screen has been deactivated")
    }
    
    func descendantOfNode(node: CCNode, indexes: UInt...) -> CCNode {
        var cur = node
        for index in indexes {
            cur = cur.children.objectAtIndex(index) as CCNode
        }
        return cur
    }

    func playSingleButtonForScreen(screen: JCSFlipUIMainMenuScreen) -> JCSButton {
        return descendantOfNode(screen, indexes: 0, 0) as JCSButton
    }
    
    func playMultiButtonForScreen(screen: JCSFlipUIMainMenuScreen) -> JCSButton {
        return descendantOfNode(screen, indexes: 0, 1) as JCSButton
    }
    
    func eventTriggered(notification: NSNotification) {
        events.append(notification)
    }
    
    class JCSFlipUIMainMenuScreenForTest : JCSFlipUIMainMenuScreen {
        
        private var status = false
        var authenticationStatus: Bool {
            get {
                return status
            }
            set {
                status = newValue
                NSNotificationCenter.defaultCenter().postNotificationName(GKPlayerAuthenticationDidChangeNotificationName, object: nil)
            }
        }
        
        override func isLocalPlayerAuthenticated() -> Bool {
            return authenticationStatus
        }
    }
}